BIP 8: Add minimum activation height

achow101 opened the pull request:
The Speedy Trial activation proposal introduces a new parameter named `minimum_activation_height`. This PR adds that to BIP 8 so that such Speedy Trials can be implemented using entirely BIP 8.

michaelfolkson commented on bip-0008.mediawiki:
```diff
@@ -50,12 +51,13 @@ The following guidelines are suggested for selecting these parameters for a soft
 # '''startheight''' should be set to some block height in the future when a majority of economic activity is expected to have upgraded to a software release including the activation parameters. Some allowance should be made for potential release delays. It should be rounded up to the next height which begins a retarget period for simplicity.
 # '''timeoutheight''' should be set to a block height when it is considered reasonable to expect the entire economy to have upgraded by, probably at least 1 year, or 52416 blocks (26 retarget intervals) after '''startheight'''.
 # '''threshold''' should be 1815 blocks (90% of 2016), or 1512 (75%) for testnet.
+# '''minimum_activation_height''' should be set to several retarget periods in the future if the startheight is to be very soon after a release. This allows more time to be spent in the locked in state so that nodes can upgrade. For a typical soft fork, this may be set to 0 to have the locked in state be the traditional single retarget period.
```
The minimum_activation_height is added primarily (imo) because the timeoutheight is short for Speedy Trial rather than because the startheight is soon after a release (though this is also true in Speedy Trial and also a factor). We would still want minimum_activation_height even if the startheight wasn't soon after a release.

achow101 replied:
My understanding is that having a longer locked in period is so that the startheight can be sooner. It allows for the upgrade time to be moved from before the start height to after locked in. Perhaps @roconnor-blockstream or @harding could comment on this.

roconnor-blockstream replied:
acho101 is correct.  I think under previous considerations a short activation would have it's start date delayed to give time for nodes to update.  Instead ST uses a miniumum_activation_height to allow for a start date soon after release while still giving time for nodes up to update.

michaelfolkson commented on bip-0008.mediawiki:
```diff
@@ -50,12 +51,13 @@ The following guidelines are suggested for selecting these parameters for a soft
 # '''startheight''' should be set to some block height in the future when a majority of economic activity is expected to have upgraded to a software release including the activation parameters. Some allowance should be made for potential release delays. It should be rounded up to the next height which begins a retarget period for simplicity.
 # '''timeoutheight''' should be set to a block height when it is considered reasonable to expect the entire economy to have upgraded by, probably at least 1 year, or 52416 blocks (26 retarget intervals) after '''startheight'''.
 # '''threshold''' should be 1815 blocks (90% of 2016), or 1512 (75%) for testnet.
+# '''minimum_activation_height''' should be set to several retarget periods in the future if the startheight is to be very soon after a release. This allows more time to be spent in the locked in state so that nodes can upgrade. For a typical soft fork, this may be set to 0 to have the locked in state be the traditional single retarget period.
```
nit: Prefer consistency across the BIP so that "LOCKED_IN" is used instead of "locked in". Two examples of "locked in" here.

achow101 replied:
Done

achow101 replied:
Done

michaelfolkson reviewed (commented):
A couple of nits but generally this looks good.

achow101 commented:
> Directly below the state transition diagram it still says "Note that when lockinontimeout is true, the LOCKED_IN state will be reached no later than at a height of timeoutheight, and **ACTIVE will be reached no later than at a height of timeoutheight + 2016**."
> 
> This needs to be changed so that it says "ACTIVE will be reached no later than at the minimum_activation_height".
> 
> minimum_activation_height still applies even when lockinontimeout is true.

Changed to "no earlier than minimum_activation_height". Also added a sentence to discuss the situation where the minimum activation height is less than the timeoutheight + 2016.

luke-jr commented on bip-0008.mediawiki:
```diff
@@ -50,12 +51,13 @@ The following guidelines are suggested for selecting these parameters for a soft
 # '''startheight''' should be set to some block height in the future when a majority of economic activity is expected to have upgraded to a software release including the activation parameters. Some allowance should be made for potential release delays. It should be rounded up to the next height which begins a retarget period for simplicity.
 # '''timeoutheight''' should be set to a block height when it is considered reasonable to expect the entire economy to have upgraded by, probably at least 1 year, or 52416 blocks (26 retarget intervals) after '''startheight'''.
 # '''threshold''' should be 1815 blocks (90% of 2016), or 1512 (75%) for testnet.
+# '''minimum_activation_height''' should be set to several retarget periods in the future if the startheight is to be very soon after a release. This allows more time to be spent in the LOCKED_IN state so that nodes can upgrade. For a typical soft fork, this may be set to 0 to have the LOCKED_IN state be the traditional single retarget period.
```
The recommendation for `startheight` should probably also be updated.

achow101 replied:
Done.

luke-jr commented on bip-0008.mediawiki:
```diff
@@ -93,7 +95,8 @@ During the MUST_SIGNAL phase, if '''(2016 - threshold)''' blocks in the retarget
 <img src="bip-0008/states.png" align="middle"></img>
 
-Note that when '''lockinontimeout''' is true, the LOCKED_IN state will be reached no later than at a height of '''timeoutheight''', and ACTIVE will be reached no later than at a height of '''timeoutheight + 2016'''.
+Note that when '''lockinontimeout''' is true, the LOCKED_IN state will be reached no later than at a height of '''timeoutheight''', and ACTIVE will be reached no earlier than at a height of '''minimum_activation_height'''.
+When '''minimum_activation_height''' is less than or equal to '''timeoutheight + 2016''', ACTIVE will be reached no later than '''timeoutheight + 2016'''.
```
The changes here aren't specific to LOT=True, but the phrasing suggests they could be.

achow101 replied:
Done.

jonasnick commented on bip-0008.mediawiki:
```diff
@@ -93,7 +95,9 @@ During the MUST_SIGNAL phase, if '''(2016 - threshold)''' blocks in the retarget
-Note that when '''lockinontimeout''' is true, the LOCKED_IN state will be reached no later than at a height of '''timeoutheight''', and ACTIVE will be reached no later than at a height of '''timeoutheight + 2016'''.
+Note that when '''lockinontimeout''' is true, the LOCKED_IN state will be reached no later than at a height of '''timeoutheight'''.
+Regardless of value of '''lockinontimeout''', ACTIVE will be reached no earlier than at a height of '''minimum_activation_height'''.
+When '''minimum_activation_height''' is less than or equal to '''timeoutheight + 2016''', ACTIVE will be reached no later than '''timeoutheight + 2016'''.
```
nit: probably (slightly) easier to understand if this wasn't special cased and instead a note would be added to the parameter selection guidelines that `minimum_activation _height` must be at least `timeoutheight + 2016`. Then you could also remove the `After at least one retarget period of LOCKED_IN [...]` and change the sentence to something like "From the LOCKED_IN state we transition to ACTIVE if the minimum activation height is reached."

jonasnick replied:
Afaics right now the diagram isn't correct, because if `minimum_activation_height = 0` then it looks like LOCKED_IN -> ACTIVE happens without a retarget period. So should either make `minimum_activation_height >= timeoutheight + 2016` or change arrow in diagram to `height >= max(timeoutheight + 2016, minimum_activation_height)`.

achow101 replied:
`minimum_activation_height` can't be at least `timeoutheight + 2016` as otherwise the normal deployment won't activate one retarget period after LOCKED_IN.

jonasnick replied:
Of course, nevermind. In that case the diagram isn't made worse by this PR.

roconnor-blockstream replied:
FWIW, I personally intended that a min-activation-height that is less than timeoutheight + 2016 would be allowed in general.  That said, my personal intentions may not be all that relevant.

ajtowns replied:
@jonasnick "it looks like LOCKED_IN -> ACTIVE happens without a retarget period" -- if that were going to happen, there'd be an arrow from "STARTED -> ACTIVE" without going through "LOCKED_IN" -- if you enter a state you're there for a retarget period, and you get to follow one of the arrows (maybe just the loop back and stay in the same state) when you hit the next retarget period.

jonasnick reviewed (commented):
Concept ACK

jonasnick reviewed (commented):
ACK

luke-jr commented on bip-0008.mediawiki:
```diff
@@ -93,7 +95,9 @@ During the MUST_SIGNAL phase, if '''(2016 - threshold)''' blocks in the retarget
-Note that when '''lockinontimeout''' is true, the LOCKED_IN state will be reached no later than at a height of '''timeoutheight''', and ACTIVE will be reached no later than at a height of '''timeoutheight + 2016'''.
+Note that when '''lockinontimeout''' is true, the LOCKED_IN state will be reached no later than at a height of '''timeoutheight'''.
+Regardless of value of '''lockinontimeout''', ACTIVE will be reached no earlier than at a height of '''minimum_activation_height'''.
+When '''minimum_activation_height''' is less than or equal to '''timeoutheight + 2016''', ACTIVE will be reached no later than '''timeoutheight + 2016'''.
```
I don't understand what this is trying to say. With LOT=False, ACTIVE might never be reached at all...

achow101 replied:
Clarified that this is after LOCKED_IN is reached.

michaelfolkson commented:
ACK

ajtowns commented on bip-0008.mediawiki:
```diff
@@ -47,15 +48,16 @@ The following guidelines are suggested for selecting these parameters for a soft
+# '''startheight''' should be set to some block height in the future. It should be rounded up to the next height which begins a retarget period for simplicity. If '''minimum_activation_height''' is not going to be set, then '''startheight''' should be set to a height when a majority of economic activity is expected to have upgraded to a software release including the activation parameters. Some a ...
 # '''timeoutheight''' should be set to a block height when it is considered reasonable to expect the entire economy to have upgraded by, probably at least 1 year, or 52416 blocks (26 retarget intervals) after '''startheight'''.
 # '''threshold''' should be 1815 blocks (90% of 2016), or 1512 (75%) for testnet.
+# '''minimum_activation_height''' should be set to several retarget periods in the future if the startheight is to be very soon after a release. This allows more time to be spent in the LOCKED_IN state so that nodes can upgrade. For a typical soft fork, this may be set to 0 to have the LOCKED_IN state be the traditional single retarget period.
```
Add triple-quotes around "startheight" for consistency?

achow101 replied:
Done

ajtowns commented on bip-0008.mediawiki:
```diff
@@ -93,7 +95,9 @@ During the MUST_SIGNAL phase, if '''(2016 - threshold)''' blocks in the retarget
-Note that when '''lockinontimeout''' is true, the LOCKED_IN state will be reached no later than at a height of '''timeoutheight''', and ACTIVE will be reached no later than at a height of '''timeoutheight + 2016'''.
+Note that when '''lockinontimeout''' is true, the LOCKED_IN state will be reached no later than at a height of '''timeoutheight'''.
+Regardless of value of '''lockinontimeout''', once LOCKED_IN, ACTIVE will be reached no earlier than at a height of '''minimum_activation_height'''.
+When '''minimum_activation_height''' is less than or equal to '''timeoutheight + 2016''', once LOCKED_IN, ACTIVE will be reached no later than '''timeoutheight + 2016'''.
```
"If LOCKED_IN is reached, the ACTIVE state will be reached either precisely 2016 blocks (one retarget period) later or at minimum_activation_height, whichever comes later." ?

achow101 replied:
Done

ajtowns commented on bip-0008/states.dot:
```diff
@@ -7,7 +7,7 @@ digraph {
   "STARTED" -> "MUST_SIGNAL" [label="height + 2016 >= timeoutheight AND lockinontimeout"];
   "STARTED" -> "FAILED" [label="height >= timeoutheight\nAND\nNOT lockinontimeout"];
-  "LOCKED_IN" -> "ACTIVE" [label="always"];
+  "LOCKED_IN" -> "ACTIVE" [label="height >= minimum_activation_height"];
```
Should be a `"LOCKED_IN":se -> "LOCKED_IN":ne;` loop edge since it can stay in LOCKED_IN for multiple periods now.

achow101 replied:
I've added the loop edge

Sjors commented:
ACK e86fd7c

luke-jr commented on bip-0008.mediawiki:
```diff
@@ -47,15 +48,16 @@ The following guidelines are suggested for selecting these parameters for a soft
 # '''name''' should be selected such that no two softforks, concurrent or otherwise, ever use the same name. For deployments described in a single BIP, it is recommended to use the name "bipN" where N is the appropriate BIP number.
 # '''bit''' should be selected such that no two concurrent softforks use the same bit. The bit chosen should not overlap with active usage (legitimately or otherwise) for other purposes.
-# '''startheight''' should be set to some block height in the future when a majority of economic activity is expected to have upgraded to a software release including the activation parameters. Some allowance should be made for potential release delays. It should be rounded up to the next height which begins a retarget period for simplicity.
+# '''startheight''' should be set to some block height in the future. It should be rounded up to the next height which begins a retarget period for simplicity. If '''minimum_activation_height''' is not going to be set, then '''startheight''' should be set to a height when a majority of economic activity is expected to have upgraded to a software release including the activation parameters. Some a ...
```
"expected software release" assumes there is only one software that needs to support softforks...

achow101 replied:
Changed to be more generic.

luke-jr commented on bip-0008.mediawiki:
```diff
@@ -47,15 +48,16 @@ The following guidelines are suggested for selecting these parameters for a soft
+# '''startheight''' should be set to some block height in the future. It should be rounded up to the next height which begins a retarget period for simplicity. If '''minimum_activation_height''' is not going to be set, then '''startheight''' should be set to a height when a majority of economic activity is expected to have upgraded to a software release including the activation parameters. Some a ...
 # '''timeoutheight''' should be set to a block height when it is considered reasonable to expect the entire economy to have upgraded by, probably at least 1 year, or 52416 blocks (26 retarget intervals) after '''startheight'''.
 # '''threshold''' should be 1815 blocks (90% of 2016), or 1512 (75%) for testnet.
+# '''minimum_activation_height''' should be set to several retarget periods in the future if the '''startheight''' is to be very soon after a release. This allows more time to be spent in the LOCKED_IN state so that nodes can upgrade. For a typical soft fork, this may be set to 0 to have the LOCKED_IN state be the traditional single retarget period.
```
I'm not sure it makes sense to talk about a "typical" or "traditional" style here. With this new param added, it may become more common to always startheight early and simply use min activation as the norm, even if it occurs before timeoutheight.

achow101 replied:
Removed that wording.

luke-jr commented on bip-0008.mediawiki:
```diff
@@ -47,15 +48,16 @@ The following guidelines are suggested for selecting these parameters for a soft
+# '''startheight''' should be set to some block height in the future. It should be rounded up to the next height which begins a retarget period for simplicity. If '''minimum_activation_height''' is not going to be set, then '''startheight''' should be set to a height when a majority of economic activity is expected to have upgraded to a software release including the activation parameters. Some a ...
 # '''timeoutheight''' should be set to a block height when it is considered reasonable to expect the entire economy to have upgraded by, probably at least 1 year, or 52416 blocks (26 retarget intervals) after '''startheight'''.
 # '''threshold''' should be 1815 blocks (90% of 2016), or 1512 (75%) for testnet.
+# '''minimum_activation_height''' should be set to several retarget periods in the future if the '''startheight''' is to be very soon after a release. This allows more time to be spent in the LOCKED_IN state so that nodes can upgrade. For a typical soft fork, this may be set to 0 to have the LOCKED_IN state be the traditional single retarget period.
```
Probably should repeat the expectation of economic majority having upgraded here.

achow101 replied:
Done
